fluent-bundle/src/scope.ts

```diff
@@ -0,0 +1,58 @@
+import { FluentBundle, FluentArgument } from "./bundle.js";
+import { FluentError } from "./error.js";
+import { RuntimeComplexPattern } from "./ast.js";
+
+export class Scope {
+  /** The bundle for which the given resolution is happening. */
+  public bundle: FluentBundle;
+  /** The list of errors collected while resolving. */
+  public errors: Array<FluentError> | null;
+  /** A dict of developer-provided variables. */
+  public args: Record<string, FluentArgument> | null;
+  /** Term references require different variable lookup logic. */
+  public insideTermReference: boolean;
+  /** The Set of patterns already encountered during this resolution.
+   * Used to detect and prevent cyclic resolutions. */
+  public dirty: WeakSet<RuntimeComplexPattern>;
+
+  constructor(
+    bundle: FluentBundle,
+    errors: Array<FluentError> | null,
+    args: Record<string, FluentArgument> | null,
+    insideTermReference = false,
+    dirty: WeakSet<RuntimeComplexPattern> = new WeakSet()
+  ) {
+    this.bundle = bundle;
+    this.errors = errors;
+    this.args = args;
+    this.insideTermReference = insideTermReference;
+    this.dirty = dirty;
+  }
+
+  cloneForTermReference(args: Record<string, FluentArgument>) {
+    return new Scope(this.bundle, this.errors, args, true, this.dirty);
+  }
+
+  reportError(error: FluentError) {
+    if (!this.errors) {
+      throw error;
+    }
+    this.errors.push(error);
+  }
+
+  memoizeIntlObject<ObjectT, OptionsT>(
```

Comment:
Assuming we keep `FluentBundle.prototype._intls` to `WeakMap<any, any>`, we may be able to drop `ObjectT` from the generics list and just use `any`.

Reply:
I like the fact that with generics here, `scope.memoizeIntlObject(Intl.PluralRules)` is correctly typed as returning an instance of `Intl.PluralRules`, so I think I'll leave it as it is right now.